okhttp/src/test/java/okhttp3/EventListenerTest.java

```diff
@@ -943,43 +948,32 @@ private void responseBodyFail(Protocol expectedProtocol) throws IOException {
         "ResponseBodyEnd", "ConnectionReleased", "CallEnd");
   }
 
-  @Test public void requestBodyFailHttp1OverHttps() throws IOException {
+  @Test public void requestBodyFailHttp1OverHttps() {
     enableTlsWithTunnel(false);
     server.setProtocols(asList(Protocol.HTTP_1_1));
-    requestBodyFail();
+
+    // Stream a 256 MiB body so the disconnect will happen before the server has read everything.
+    requestBodyFail(new MockResponse()
+        .setSocketPolicy(SocketPolicy.DISCONNECT_DURING_REQUEST_BODY), sizedRequest(256), Protocol.HTTP_1_1);
   }
 
-  @Test public void requestBodyFailHttp2OverHttps() throws IOException {
+  @Test public void requestBodyFailHttp2OverHttps() {
     enableTlsWithTunnel(false);
     server.setProtocols(asList(Protocol.HTTP_2, Protocol.HTTP_1_1));
-    requestBodyFail();
-  }
 
-  @Test public void requestBodyFailHttp() throws IOException {
-    requestBodyFail();
+    requestBodyFail(new MockResponse()
+        .setSocketPolicy(SocketPolicy.DISCONNECT_DURING_REQUEST_BODY), sizedRequest(2), Protocol.HTTP_2);
```

Comment:
I'm worried about this, it might be better to have a custom request body that throttles the writes as well...